paramstorer.go

```diff
@@ -67,9 +64,82 @@ type mapParamStorer struct {
 }
 
 func (m *mapParamStorer) SafeParams() map[string]interface{} {
+	if m.safeParams == nil {
+		return map[string]interface{}{}
+	}
 	return m.safeParams
 }
 
 func (m *mapParamStorer) UnsafeParams() map[string]interface{} {
+	if m.unsafeParams == nil {
+		return map[string]interface{}{}
+	}
 	return m.unsafeParams
 }
+
+func (m *mapParamStorer) putSafeParam(k string, v interface{}) {
+	if m.safeParams == nil {
+		m.safeParams = map[string]interface{}{k: v}
+	} else {
+		m.safeParams[k] = v
+	}
+	delete(m.unsafeParams, k)
+}
+
+func (m *mapParamStorer) putUnsafeParam(k string, v interface{}) {
+	if m.unsafeParams == nil {
+		m.unsafeParams = map[string]interface{}{k: v}
+	} else {
+		m.unsafeParams[k] = v
+	}
+	delete(m.safeParams, k)
+}
+
+func (m *mapParamStorer) copyFrom(storer ParamStorer) {
+	if storer == nil {
+		return
+	}
+	// If this is one of our types, we can access the values directly and avoid intermediate map allocations.
+	switch st := storer.(type) {
+	case singleParamStorer:
+		if st.safe {
+			m.putSafeParam(st.key, st.value)
+		} else {
+			m.putUnsafeParam(st.key, st.value)
+		}
+	case *mapParamStorer:
+		for k, v := range st.safeParams {
```

Comment:
note: I wonder if we get more savings if we use the index rather than loop variable on all of these loops. I imagine the length of each one of these maps is quite low, so it might just hurt readability for the cost of minimal improvements, but just a thought.

Reply:
I can't explain exactly why, but I ran the benchmarks and it's actually slower. I'm guessing there are optimizations in the for loop for pulling the key and value at the same time, while accessing the map value directly requires a second lookup in the hash table.

Reply:
Interesting 😕 thanks for checking it out